File: src/trades/account.cpp
#include "trade\tdx.h"
#include "cube\safe.h"
#include "cube\log\log.h"
#include "trades\config.h"
#include "trades\account.h"
BEGIN_TRADES_NAMESPACE
accounts *accounts::_instance = 0;

accounts *accounts::instance() {
	if (_instance == 0) {
		_instance = new accounts();
	}

	return _instance;
}

int accounts::login(const std::string &laccount, const std::string taccount, const std::string tpwd, const std::string cpwd, const std::string &ip, ushort port, int dept, const std::string &version, std::string *error) {
	cube::log::info("account %s login...", laccount.c_str());
	std::lock_guard<std::mutex> lck(_mutex);
	//account has login
	if (_accounts.find(laccount) != _accounts.end()) {
		cube::log::info("account %s has already login.", laccount.c_str());
		return 0;
	}

	std::string errmsg("");
	//account create
	trade::trade *acnt = new trade::tdx();
	if (acnt->init(config::wdir, &errmsg) != 0) {
		cube::log::error("account %s login failed, error: %s", laccount.c_str(), errmsg.c_str());
		cube::safe_assign<std::string>(error, errmsg);
		return -1;
	}

	//account login
	if (acnt->login(ip, port, version, dept, laccount, taccount, tpwd, cpwd, &errmsg) != 0) {
		cube::safe_assign<std::string>(error, errmsg);
		cube::log::error("account %s login failed, error: %s", laccount.c_str(), errmsg.c_str());
		return -1;
	}

	//add account
	_accounts.insert(std::pair<std::string, trade::trade*>(laccount, acnt));

	cube::log::info("account %s login success.", laccount.c_str());
	return 0;

}

int accounts::quote(const std::string &account, const std::string &code, trade::table &result, std::string *error) {
	cube::log::info("query quote data with code %s from account %s...", code.c_str(), account.c_str());
	std::lock_guard<std::mutex> lck(_mutex);
	std::map<std::string, trade::trade*>::iterator iter = _accounts.find(account);
	//find account
	if (iter == _accounts.end()) {
		cube::safe_assign<std::string>(error, "account not exist.");
		cube::log::error("query quote data with code %s from account %s failed, error: account not exist", code.c_str(), account.c_str());
		return -1;
	}

	std::string errmsg("");
	//query data
	if (iter->second->quote(code, result, &errmsg) != 0) {
		cube::safe_assign<std::string>(error, errmsg);
		cube::log::error("query quote data with code %s from account %s failed, error: %s", code.c_str(), account.c_str(), errmsg.c_str());
		return -1;
	}

	cube::log::info("query quote data with code %s from account %s success.", code.c_str(), account.c_str());
	return 0;
}

int accounts::query(const std::string &account, int category, trade::table &result, std::string *error) {
	cube::log::info("query account %s data with category %d ...", account.c_str(), category);
	std::lock_guard<std::mutex> lck(_mutex);
	std::map<std::string, trade::trade*>::iterator iter = _accounts.find(account);
	//find account
	if (iter == _accounts.end()) {
		cube::safe_assign<std::string>(error, "account not exist.");
		cube::log::error("query account %s data with category %d failed, error: account not exist", account.c_str(), category);
		return -1;
	}

	std::string errmsg("");
	//query data
	if (iter->second->query((trade::query::type)category, result, &errmsg) != 0) {
		cube::safe_assign<std::string>(error, errmsg);
		cube::log::error("query account %s data with category %d failed, error: %s", account.c_str(), category, errmsg.c_str());
		return -1;
	}

	cube::log::info("query account %s data with category %d success.", account.c_str(), category);
	return 0;
}

int accounts::query(const std::string &account, int category, const std::string &start_date, const std::string &end_date, trade::table &result, std::string *error) {
	cube::log::info("query account %s history data from %s~%s with category %d ...", account.c_str(), start_date.c_str(), end_date.c_str(), category);
	std::lock_guard<std::mutex> lck(_mutex);
	std::map<std::string, trade::trade*>::iterator iter = _accounts.find(account);
	//find account
	if (iter == _accounts.end()) {
		cube::safe_assign<std::string>(error, "account not exist.");
		cube::log::error("query account %s history data from %s~%s with category %d failed, error: account not exist", account.c_str(), start_date.c_str(), end_date.c_str(), category);
		return -1;
	}

	std::string errmsg("");
	//query data
	if (iter->second->query((trade::query::type)category, start_date, end_date, result, &errmsg) != 0) {
		cube::safe_assign<std::string>(error, errmsg);
		cube::log::error("query account %s history data from %s~%s with category %d failed, error: %s", account.c_str(), start_date.c_str(), end_date.c_str(), category, errmsg.c_str());
		return -1;
	}

	cube::log::info("query account %s history data from %s~%s with category %d success.", account.c_str(), start_date.c_str(), end_date.c_str(), category);
	return 0;
}

int accounts::order(const std::string &account, int otype, int ptype, const std::string &gddm, const std::string &zqdm, float price, int count, trade::table &result, std::string *error) {
	cube::log::info("order <otype:%d, ptype:%d, gddm:%s, zqdm:%s, price: %.2f, count: %d> from account %s...", otype, ptype, gddm.c_str(), zqdm.c_str(), price, count, account.c_str());
	std::lock_guard<std::mutex> lck(_mutex);
	std::map<std::string, trade::trade*>::iterator iter = _accounts.find(account);
	//find account
	if (iter == _accounts.end()) {
		cube::safe_assign<std::string>(error, "account not exist.");
		cube::log::error("order <otype:%d, ptype:%d, gddm:%s, zqdm:%s, price: %.2f, count: %d> from account %s failed, error: account not exist", otype, ptype, gddm.c_str(), zqdm.c_str(), price, count, account.c_str());
		return -1;
	}

	std::string errmsg("");
	//send order
	if (iter->second->send((trade::order::type)otype, (trade::price::type)ptype, gddm, zqdm, price, count, result, &errmsg) != 0) {
		cube::safe_assign<std::string>(error, errmsg);
		cube::log::info("order <otype:%d, ptype:%d, gddm:%s, zqdm:%s, price: %.2f, count: %d> from account %s failed. error: %s", otype, ptype, gddm.c_str(), zqdm.c_str(), price, count, account.c_str(), errmsg.c_str());
		return -1;
	}

	cube::log::info("order <otype:%d, ptype:%d, gddm:%s, zqdm:%s, price: %.2f, count: %d> from account %s success", otype, ptype, gddm.c_str(), zqdm.c_str(), price, count, account.c_str());
	return 0;
}

int accounts::cancel(const std::string &account, const std::string &seid, const std::string &orderno, trade::table &result, std::string *error) {
	cube::log::info("cancel <seid:%s, orderno:%s from account %s...", seid.c_str(), orderno.c_str(), account.c_str());
	std::lock_guard<std::mutex> lck(_mutex);
	std::map<std::string, trade::trade*>::iterator iter = _accounts.find(account);
	//find account
	if (iter == _accounts.end()) {
		cube::safe_assign<std::string>(error, "account not exist.");
		cube::log::info("cancel <seid:%s, orderno:%s from account %s failed, error: account not exist", seid.c_str(), orderno.c_str(), account.c_str());
		return -1;
	}

	std::string errmsg("");
	//cancel order
	if (iter->second->cancel(seid, orderno, result, &errmsg) != 0) {
		cube::safe_assign<std::string>(error, errmsg);
		cube::log::info("cancel <seid:%s, orderno:%s from account %s failed, error: %s", seid.c_str(), orderno.c_str(), account.c_str(), errmsg.c_str());
		return -1;
	}

	cube::log::info("cancel <seid:%s, orderno:%s from account %s success", seid.c_str(), orderno.c_str(), account.c_str());
	return 0;
}

int accounts::logout(const std::string &account, std::string *error) {
	std::lock_guard<std::mutex> lck(_mutex);
	std::map<std::string, trade::trade*>::iterator iter = _accounts.find(account);
	//account has login
	if (iter == _accounts.end()) {
		cube::safe_assign<std::string>(error, "account not exist.");
		cube::log::error("account %s logout failed, error: account not exist.", account.c_str());
		return -1;
	}

	//logout
	iter->second->logout();

	//destroy
	iter->second->destroy();

	//delete
	delete iter->second;

	//remove
	_accounts.erase(iter);

	return 0;
}
END_TRADES_NAMESPACE

File: app/svc/tlib/stock/trade/tdx/account.py
"""
    securities account for trade
"""
from .. import account, error
from . import trader, protocol, util


class Account(account.Account):
    def __init__(self, **kwargs):
        """
            init account
        :param laccount: str, 登录账户，一般为资金账户或者客户好
        :param lpwd: str, 登录密码
        :param taccount: str, 交易账户，一般为资金账户
        :param tpwd: str, 交易密码
        :param dept: str, 营业部代码
        :param version: str, 客户端版本号
        :param agentservers: array, 代理服务器地址列表, [(id, host, port), (id, host, port), ...]
        :param tradeservers: array, 券商交易服务器地址列表, [(id, ip, port), (id, ip, port), ...]
        """
        # get args
        acnt, pwd = kwargs.get('account'), kwargs.get('pwd')
        laccount, lpwd, taccount, tpwd = kwargs.get('laccount', acnt), kwargs.get('lpwd', pwd), kwargs.get('taccount', acnt), kwargs.get('tpwd', pwd)
        dept, version, gddm, agents, servers = kwargs.get('dept'), kwargs.get('version'), kwargs.get('gddm'), kwargs.get('agents'), kwargs.get('servers')

        # check args
        if laccount is None or lpwd is None or taccount is None or tpwd is None or dept is None or version is None or gddm is None or agents is None or servers is None:
            raise error.TradeError('missing account parameters')

        # init account & traders
        self._account = trader.Account(laccount, lpwd, taccount, tpwd, dept, version, gddm)
        self._traders = trader.Traders(self._account, agents, servers)

    def login(self):
        """
            账户登录
        :return: dict
            {'status': status, 'msg': msg, 'data':{}}
        """
        try:
            return self._traders.login()
        except Exception as e:
            return protocol.failed(str(e))

    def logout(self):
        """
            账户退出
        :return:
            (True|False, result message string)
        """
        try:
            return self._traders.logout()
        except Exception as e:
            return protocol.failed(str(e))

    def status(self):
        """
            get account status
        :return:
        """
        try:
            data = self._account.status()
            data['traders'] = self._traders.status()
            return protocol.success(data=data)
        except Exception as e:
            return protocol.failed(str(e))

    def quote(self, zqdm):
        """
            查询股票实时行情
        :param zqdm: str, stock code
        :return:
            dict
        """
        try:
            # send query
            return self._traders.quote(zqdm)
        except Exception as e:
            return protocol.failed(str(e))

    def query(self, type, sdate=None, edate=None):
        """
            查询账户当前或者历史数据
        :param type: str, data type: dqzc/dqcc/drwt/drcj/kcwt/gdxx/lswt/lscj/jgd
        :param sdate: str, in, format: yyyymmdd
        :param edate: str, in, format: yyyymmdd
        :return:
            list/dict
        """
        try:
            # send query
            if type in protocol.const.query['current'].keys():
                return self._traders.queryc(protocol.const.query['current'][type])
            elif type in protocol.const.query['history'].keys():
                return self._traders.queryh(protocol.const.query['history'][type], sdate, edate)
            else:
                return protocol.failed('非法查询类型:%s'%type)
        except Exception as e:
            return protocol.failed(str(e))


    def place(self, otype, ptype, zqdm, price, count):
        """
            委托下单
        :param otype: order type: buy/sell
        :param ptype: price type: xj/sj
        :param zqdm: stock code
        :param price: order price
        :param count: order count
        :return:
        """
        try:
            # check order type
            if otype not in protocol.const.otype.keys():
                return protocol.failed('非法的委托方向: %s'%otype)
            # check price type
            if ptype not in protocol.const.ptype.keys():
                return protocol.failed(('非常的报价类型: %s'%ptype))

            # send order to remote
            return self._traders.order(protocol.const.otype[otype], protocol.const.ptype[ptype], zqdm, price, count)
        except Exception as e:
            return protocol.failed(str(e))

    def cancel(self, zqdm, orderno):
        """
            委托撤单
        :param zqdm: str, stock code
        :param orderno: str, order number
        :return:
        """
        try:
            # get se id
            seid = util.getse(zqdm)
            if seid is None:
                return protocol.failed('非法的证券代码: %s'%zqdm)

            # cancel order
            return self._traders.cancel(seid, orderno)
        except Exception as e:
            return protocol.failed(str(e))

# register channel
account.register('tdx', Account)
